### DISCUSSION TITLE
Using httpx-ws as background task to test websockets

### DISCUSSION THREAD

**Discussion by wholmen** [frankie567/httpx-ws#13] [Q&A]
Hi, reposting the issue to this github.  

I am struggling with httpx-ws. It works when the server-side websocket connection is closed, but my implementations are based on websockets that are open until the client breaks them. Could you help me adjusting the code to make the implementation work? 

### Websocket implementation
```python
from fastapi import APIRouter, Depends, WebSocket, WebSocketDisconnect
import asyncio

@router.websocket("/ws")
async def websocket_labtest(websocket: WebSocket):
    await websocket.accept()
    sub_id: uuid.UUID = uuid.uuid4()
    subjects.ADD_LABTEST_QUEUE[sub_id] = asyncio.Queue()

    try:
        while True:
            event: events.LabtestAddedEvent = await subjects.ADD_LABTEST_QUEUE[sub_id].get()
            await websocket.send_json(event.json())
    except WebSocketDisconnect:
        subjects.ADD_LABTEST_QUEUE.pop(sub_id)
```

### Test implementation
```python
@pytest.fixture
async def labtest_client_async() -> AsyncClient:
    async with AsyncClient(transport=ASGIWebSocketTransport(app), base_url='http://test') as client:
        yield client


async def subscribe_to_labtests_ws_async(labtest_client_async: httpx.AsyncClient, queue: asyncio.Queue):
    async with httpx_ws.aconnect_ws("/labtests/ws", labtest_client_async) as ws:
        message = await ws.receive_json()
        queue.put_nowait(message)


@pytest.mark.asyncio
async def test_add_lab_test(labtest_client_async: httpx.AsyncClient):
    sub_queue: asyncio.Queue = asyncio.Queue()
    task = asyncio.create_task(subscribe_to_labtests_ws_async(labtest_client_async, sub_queue))

    try:
        await asyncio.sleep(SHORT_WAIT)
        await add_labtest(labtest_client_async)
        await asyncio.sleep(SHORT_WAIT)
    finally:
        task.cancel()

    message_labtest1 = await sub_queue.get()
    print(f"{message_labtest1=}")
```

The websocket works fine when I use e.g. Postman to connect to the websocket. It seems httpx_ws is unable to close the connection even though I say `task.cancel()`. That makes the test run until I force quit. 

It seems like the problem being related to `asyncio.create_task`. If I don't wait for the "add labtest event" in server, I manage to get the message sent to me. So it seems like the connection struggles to run as a background task. Maybe this is completely out of scope. I am open to ideas on how to do this kind of integration testing. 

I know the code is not self-contained. If it's hard to understand, I can make a simpler self-contained example

**Comment by frankie567** [comment]
Hi @wholmen 👋 

Is there a reason why you want to run your client connection into a task? IMO, it would be way easier for debugging/reasoning to just run it in the main context:

```py
@pytest.mark.asyncio
async def test_add_lab_test(labtest_client_async: httpx.AsyncClient):
    sub_queue: asyncio.Queue = asyncio.Queue()
    async with httpx_ws.aconnect_ws("/labtests/ws", labtest_client_async) as ws:
        message = await ws.receive_json()
        await sub_queue.put(message)
        await asyncio.sleep(SHORT_WAIT)

    message_labtest1 = await sub_queue.get()
    print(f"{message_labtest1=}")
```

**Reply by wholmen** [reply]
Yes, I need it to be a background task, because it's the `await add_labtest(labtest_client_async)` that tells the websocket client to fire a message to the client. 

So the full end-to-end system I want to test here is: 

1. (client) Start subscription to a webservice
2. (client) Send rest api call to add a lab test to the database
3. (server) adds labtest to database
4. (server) Server puts an event to a queue that will tell every subscriber that a labtest was added
5. (server) Websocket picks up the event and notifies every webservice subscribers 
6. (client) Picks up the message that the labtest was created
7. (client) Query API to retrieve data on the new labtest

I followed this pattern with graphql and subscriptions, but I wanted to get rid of graphql for other reasons. So now I'm trying to replace subscriptions with websocket, but it proves a bit difficult to test it with pytest. 

The information flow I have described works with manual testing. I'm just struggling to test it with pytest. 

I could of course just skip end-to-end-testing, but I like to ensure these more advanced patterns do not break

**Reply by wholmen** [reply]
I could write a more complete (self-contained) example if the code example I provided is not obvious

**Reply by frankie567** [reply]
Ok, I see. I think the problem here is that you're using the same HTTPX client instance. I suspect the first one is blocked by the pending WebSocket Client. Could you try to open another one to make your `add_labtest` request?

**Reply by wholmen** [reply]
I tried to use a different httpx AsyncClient now for the post request. 

The background task still doesn't manage to pick up the websocket message, and it doesn't manage to shut down after the test has passed

**Reply by frankie567** [reply]
If you could provide a simplified example, that would be great and help me to understand all the bits and pieces.

**Comment by wholmen** [comment]
Ok, I have been a little busy, but I have created a full example showing the workflow. I have tested this with Postman, so it should in principle be testable. 

# Implementation of websocket
```python
import asyncio
import uuid
from typing import Dict

from fastapi import APIRouter, Response, status, WebSocket, WebSocketDisconnect


router = APIRouter(tags=["Test e2e of websocket"])

SOCKET_QUEUE: Dict[uuid.UUID, asyncio.Queue[int]] = {}


@router.post("/activate_websocket/{post_id}")
def activate_websocket(post_id: int):
    for socket in SOCKET_QUEUE.keys():
        SOCKET_QUEUE[socket].put_nowait(post_id)

    return Response(status_code=status.HTTP_200_OK)


@router.websocket_route("/test_ws")
async def ws_test(websocket: WebSocket):
    await websocket.accept()
    socket_id: uuid.UUID = uuid.uuid4()
    SOCKET_QUEUE[socket_id] = asyncio.Queue()

    try:
        while True:
            post_id: int = await SOCKET_QUEUE[socket_id].get()
            await websocket.send_text(f"Post id {post_id} was sent to active_websocket")
    except WebSocketDisconnect:
        SOCKET_QUEUE.pop(socket_id)
```

# Test
```python
import pytest
import asyncio
import httpx
import httpx_ws
from httpx import AsyncClient
from httpx_ws.transport import ASGIWebSocketTransport
from fastapi import status

SHORT_WAIT = 0.2


@pytest.fixture
async def labtest_client_async() -> AsyncClient:
    async with AsyncClient(transport=ASGIWebSocketTransport(app), base_url="http://test") as client:
        yield client


async def subscribe_to_simple_websocket_async(labtest_client_async: httpx.AsyncClient, queue: asyncio.Queue):
    async with httpx_ws.aconnect_ws("http://test/test_ws", labtest_client_async) as ws:
        message = await ws.receive_text()
        queue.put_nowait(message)


async def post_id_to_send_websocket_message(labtest_client_async: httpx.AsyncClient, post_id: int):
    url = f"/activate_websocket/{post_id}"
    response: httpx.Response = await labtest_client_async.post(url)
    assert response.status_code == status.HTTP_200_OK


@pytest.mark.asyncio
async def test_listen_to_websocket(labtest_client_async: httpx.AsyncClient):
    sub_queue: asyncio.Queue = asyncio.Queue()
    task = asyncio.create_task(subscribe_to_simple_websocket_async(labtest_client_async, sub_queue))

    try:
        await asyncio.sleep(SHORT_WAIT)
        await post_id_to_send_websocket_message(labtest_client_async, post_id=2)
        await asyncio.sleep(SHORT_WAIT)
        await post_id_to_send_websocket_message(labtest_client_async, post_id=3)
        await asyncio.sleep(SHORT_WAIT)
    finally:
        task.cancel()

    message_first_post_id1 = await sub_queue.get()
    print(f"{message_first_post_id1=}")
    message_first_post_id2 = await sub_queue.get()
    print(f"{message_first_post_id2=}")
```

Testing this from Postman works fine, and every post-request I send will trigger the websocket: 
![image](https://user-images.githubusercontent.com/5293373/205926800-721480fa-ea26-4793-ba6b-9fd4cddba02d.png)

The reason I want this pattern, is that I want to have items posted to the database from rabbitmq, and then notify the client when it happens. I also want to be able to parse items one by one, and tell the client that things are ready. I am making software that has some heavy computational components

edit: added two imports to make fixture work. It's in conftest.py in my setup, so I made a mistake when copying it here

**Reply by wholmen** [reply]
This exact test-pattern worked fine when testing strawberry-graphql subscriptions, but I might not use the correct async-patterns. 

If you have ideas for changing the pattern, please tell me

**Reply by frankie567** [reply]
Thanks, I'll have a look into that.

One question that comes to my mind: you say that you use RabbitMQ as a message broker to notify from events. Wouldn't it be simpler to actually subscribe to the queue when the WebSocket is opened? Something like (pseudo-code):

```py
async def ws_endpoint(ws: WebSocket):
    with rabbitmq_client.subscribe("events") as subscription:
        while True:
            event = subscription.wait()
            await ws.send_text("New event", event)
```

In this pattern one WebSocket subscription == one RabbitMQ subscription. You don't need to bother with HTTP calls, all is managed by the message broker.

**Reply by wholmen** [reply]
Yes, that might be a good idea for the rabbitmq-part of it, and I'll definitively see if that can be solved. I would guess this is harder to test, and it's also makes the code more technology-dependent. Having an internal async-queue-method might be a nice abstraction. Or a unnecessary complication. I will test it out 

But, it doesn't solve my other problem. Which is not really displayed in this code. 

I will have to e.g. generate 40-50 reports, some of them takes up to a minute to produce because of huge datasets. Then I would like to have my client fire off an post-request and use a websocket to notify every time a new report is ready. I think I would need this pattern here. Or, I could of course fire off events to myself via rabbit, but it seems weird that one would need an external queue system for a simple pattern like that.

**Reply by wholmen** [reply]
The main problem here is the automatic e2e testing of this pattern. Not the implementation or use by the client. 

This works, but I can't figure out how to test it :)

**Reply by frankie567** [reply]
Ok, so I achieved to make it work (kinda). In your server code, you used `asyncio.Queue` to keep track of opened WebSockets. The problem is that `ASGIWebSocketTransport` starts its own thread to run the ASGI app. Thus, they don't run on the same event loop, so the queues won't be shared.

By changing it to thread-safe `queue.Queue`, we can experience the desired behavior:

```py
import queue
import uuid
from typing import Dict

from fastapi import APIRouter, Response, status, WebSocket, WebSocketDisconnect, FastAPI


router = APIRouter(tags=["Test e2e of websocket"])

SOCKET_QUEUE: Dict[uuid.UUID, queue.Queue[int]] = {}


@router.post("/activate_websocket/{post_id}")
async def activate_websocket(post_id: int):
    print("SOCKET_QUEUE", SOCKET_QUEUE)
    for socket in SOCKET_QUEUE.keys():
        SOCKET_QUEUE[socket].put(post_id)

    return Response(status_code=status.HTTP_200_OK)


@router.websocket_route("/test_ws")
async def ws_test(websocket: WebSocket):
    await websocket.accept()
    socket_id: uuid.UUID = uuid.uuid4()
    SOCKET_QUEUE[socket_id] = queue.Queue()
    print("SOCKET_QUEUE", SOCKET_QUEUE)

    try:
        while True:
            post_id: int = SOCKET_QUEUE[socket_id].get()
            await websocket.send_text(f"Post id {post_id} was sent to active_websocket")
    except WebSocketDisconnect:
        SOCKET_QUEUE.pop(socket_id)
```

**However**, while it'll solve your specific problem here, this approach won't work in the real-world. Indeed, web servers usually spawn several processes to process requests (or, we can imagine a multi-servers deployment). In this case, you can't be sure that your POST request and the WebSocket request will be processed by the same process. If they're not, they won't share the same memory queues, so they won't be able to talk to each other.

That's why it's important to rely on a central broker such as RabbitMQ if you need your web server and background worker to communicate.

From what I understand of what you want to do, I would imagine the following architecture:

```mermaid
flowchart LR
    CLIENT((Client))
    WORKER[Worker]
    BROKER[(RabbitMQ)]
    SERVER[Web server]
    CLIENT -- "(HTTP) asks for a report" --> SERVER
    SERVER -- "enqueue a todo report" --> BROKER
    BROKER -- "notifies a todo report" --> WORKER
    WORKER -- "report finished" --> BROKER
    BROKER -- "report finished" --> SERVER
    SERVER -- "(WS) notifies report is finished" --> CLIENT
```

The HTTP request will just enqueue a new report job in the broker, that will be picked by the worker. You can return a unique ID for this job, so your client can identify it when getting the notification.

The WebSocket, on the other hand, just has to subscribe to the broker and wait for "report finished" event, and send back the job ID to the client.

In this configuration, testing is easier, because you can test in isolation if a job is correctly sent when making the HTTP request and if the WS correctly forwards the event when a job is finished. We could very well imagine an interface for the broker pub/sub, that could be mocked to ease testing.

**Reply by wholmen** [reply]
Thanks a lot for the explanation, @frankie567! It makes a lot of sense. I didn't think about multiprocessing/multi-pods, so yeah, that makes it necessary to have a broker for this communication. 

I didn't get the new code to work either. Did you update the test to use queue instead of asyncio as well?

**Reply by frankie567** [reply]
You're welcome!

Hmm, apart from adding some `print` there and there, I don't think so:

```py
import pytest
import pytest_asyncio
import asyncio
import httpx
import httpx_ws
from httpx_ws.transport import ASGIWebSocketTransport
from fastapi import status

from .ws_server import app

SHORT_WAIT = 0.2


@pytest_asyncio.fixture
async def labtest_client_async() -> httpx.AsyncClient:
    async with httpx.AsyncClient(transport=ASGIWebSocketTransport(app), base_url="http://test") as client:
        yield client


async def subscribe_to_simple_websocket_async(labtest_client_async: httpx.AsyncClient, queue: asyncio.Queue):
    print("Hello, I'm starting")
    async with httpx_ws.aconnect_ws("http://test/test_ws", labtest_client_async) as ws:
        while True:
            print("WS Receive...")
            message = await ws.receive_text()
            print("WS Received", message)
            await queue.put(message)


async def post_id_to_send_websocket_message(labtest_client_async: httpx.AsyncClient, post_id: int):
    url = f"/activate_websocket/{post_id}"
    print("HTTP POST...")
    response: httpx.Response = await labtest_client_async.post(url)
    print("HTTP POSTED!")
    assert response.status_code == status.HTTP_200_OK


@pytest.mark.asyncio
async def test_listen_to_websocket(labtest_client_async: httpx.AsyncClient):
    sub_queue: asyncio.Queue = asyncio.Queue()
    task = asyncio.create_task(subscribe_to_simple_websocket_async(labtest_client_async, sub_queue))

    try:
        await post_id_to_send_websocket_message(labtest_client_async, post_id=2)
        await asyncio.sleep(SHORT_WAIT)
        await post_id_to_send_websocket_message(labtest_client_async, post_id=2)
        await asyncio.sleep(SHORT_WAIT)
        await post_id_to_send_websocket_message(labtest_client_async, post_id=3)
        await asyncio.sleep(SHORT_WAIT)
    finally:
        task.cancel()

    message_first_post_id1 = await sub_queue.get()
    print(f"{message_first_post_id1=}")
    message_first_post_id2 = await sub_queue.get()
    print(f"{message_first_post_id2=}")
```

**Reply by wholmen** [reply]
You added `while True:` to the websocket test. That's interesting. Didn't think of that, but it makes sense. Works now :) 

Thanks a lot!

**Reply by wholmen** [reply]
The test don't end properly, but I can look into that myself. Since I might not end up using this pattern after all, you don't have to spend much more time on it :)

**Reply by frankie567** [reply]
Oh yeah, right indeed! Since you want to receive several messages, the `while` loop is necessary. And yes, the test doesn't end up properly. You would likely need a mechanism to signal the thread to stop, e.g. using [`threading.Event`](https://docs.python.org/3/library/threading.html#threading.Event).

**Reply by wholmen** [reply]
Is it possible to close the websocket from the client side using httpx_ws? 

A brute force `ws.close()` seems to do nothing. The background task is ended properly, but none of them does the trick.

**Reply by frankie567** [reply]
Exiting the context manager is enough to close the client connection, it's implicitly closed when leaving the context manager.

The problem that may happen is that you're stuck waiting for `receive_text`, so you don't have the opportunity to go back to the start of the loop. You can add a timeout to prevent it from blocking for too long: `await ws.receive_text(timeout=0.1)`
